src/algorithms/synthesis/syrec_synthesis_additional_lines.cpp

```diff
@@ -0,0 +1,126 @@
+                bitwise_cnot(lines, lhs); // duplicate lhs
+                bitwise_cnot(lines, rhs);
+                break;
+            case binary_expression::multiply: // *
+                get_constant_lines(expression.bitwidth(), 0u, lines);
+                multiplication(lines, lhs, rhs);
+                break;
+            case binary_expression::divide: // /
+                get_constant_lines(expression.bitwidth(), 0u, lines);
+                division(lines, lhs, rhs);
+                break;
+            case binary_expression::modulo: { // %
+                get_constant_lines(expression.bitwidth(), 0u, lines);
+                std::vector<unsigned> quot;
+                get_constant_lines(expression.bitwidth(), 0u, quot);
+
+                bitwise_cnot(lines, lhs); // duplicate lhs
+                modulo(quot, lines, rhs);
+            } break;
+            case binary_expression::logical_and: // &&
+                lines.emplace_back(get_constant_line(false));
+                conjunction(lines.at(0), lhs.at(0), rhs.at(0));
+                break;
+            case binary_expression::logical_or: // ||
+                lines.emplace_back(get_constant_line(false));
+                disjunction(lines.at(0), lhs.at(0), rhs.at(0));
+                break;
+            case binary_expression::bitwise_and: // &
+                get_constant_lines(expression.bitwidth(), 0u, lines);
+                bitwise_and(lines, lhs, rhs);
+                break;
+            case binary_expression::bitwise_or: // |
+                get_constant_lines(expression.bitwidth(), 0u, lines);
+                bitwise_or(lines, lhs, rhs);
+                break;
+            case binary_expression::less_than: // <
+                lines.emplace_back(get_constant_line(false));
+                less_than(lines.at(0), lhs, rhs);
+                break;
+            case binary_expression::greater_than: // >
+                lines.emplace_back(get_constant_line(false));
+                greater_than(lines.at(0), lhs, rhs);
+                break;
+            case binary_expression::equals: // =
+                lines.emplace_back(get_constant_line(false));
+                equals(lines.at(0), lhs, rhs);
+                break;
+            case binary_expression::not_equals: // !=
+                lines.emplace_back(get_constant_line(false));
+                not_equals(lines.at(0), lhs, rhs);
+                break;
+            case binary_expression::less_equals: // <=
+                lines.emplace_back(get_constant_line(false));
+                less_equals(lines.at(0), lhs, rhs);
+                break;
+            case binary_expression::greater_equals: // >=
+                lines.emplace_back(get_constant_line(false));
+                greater_equals(lines.at(0), lhs, rhs);
+                break;
+            default:
+                return false;
+        }
+
+        return true;
+    }
```

Comment:
There still is a lot of code duplication between this function and the implementation in the other class.
However, the differences in both are very specific and I did not have the time to refactor these details so that the implementations can be unified.

One possible solution would be to define virtual methods in the base class for all the cases where the synthesisers differ and appropriately overload them in the derived classes. As a result, this function can be moved to the base class as well.

Reply:
done 👍🏼